### PR TITLE
[docs] Document the `browser-init-timeout` feature

### LINKED ISSUE #5861: [docs] document the browser initialization timeout
While working on #5720, we've found that it could be helpful for the user to be able to modify the default timeout during which a browser should establish a connection. Before the changes were made, the timeout was equal to 2 minutes if there is at least one local browser among the launched browsers, and 6 minutes otherwise (same timeout for all browsers).

With introducing a way to modify the timeout value via "--browser-init-timeout" (and "browserInitTimeout" option of the [runner.run](https://devexpress.github.io/testcafe/documentation/reference/testcafe-api/runner/run.html) method and configuration file), one can now manually set the time during which a browser should establish a connection. When timeout is set manually by the user, it will override the default logic (2 mins for local browsers and 6 mins for remotes) and all types of browsers will have the same timeout value.

### PR DESCRIPTION
**Pull request by DIRECTcut**
fixes #5861

### CODE REVIEW

**Comment by DIRECTcut** [comment]
This timeout has 2 possible default values (they differ for remote and local browsers). I am unsure which of the following looks better:

<details>
  <summary>Bullet list with a bold heading</summary>
  
image
  
</details>

<details>
  <summary>Bullet list with bold items and no heading</summary>
  
image
  
</details>

**Comment by DIRECTcut** [comment]
The defaults of a runner option are also quastionable:

<details>
  <summary>Bullet list with a bold heading</summary>
  
image
  
</details>

**Comment by DIRECTcut** [comment]
During my work on this issue, I've noticed how we mention API methods in the [config file description](https://devexpress.github.io/testcafe/documentation/reference/configuration-file.html#pageloadtimeout).

For booleans it seems fine since the presence of a parameter means it's set to `true`:

image

We do the same forparameters that take numbers (for instance, [pageloadtimeout]((https://devexpress.github.io/testcafe/documentation/reference/configuration-file.html#pageloadtimeout)):

image

Shouldn't we instead do something like:

> API: runner.run({ pageLoadTimeout: 2000 })

**Comment by VasilyStrelyaev** [comment]
> This timeout has 2 possible default values (they differ for remote and local browsers). I am unsure which of the following looks better:

The first looks much better, I agree.

> The defaults of a runner option are also questionable:

They look okay this way.

> During my work on this issue, I've noticed how we mention API methods in the config file description.

BTW, the presence of a boolean parameter doesn't seem to mean it's `true`...

Technically, there's no issue with this indication IMO.

```js
const stopOnFirstFail = true;
runner.run({ stopOnFirstFail });
```

```js
const pageLoadTimeout = 2000;
runner.run({ pageLoadTimeout });
```

This works for both properties.

I don't think we should make links (!) even longer by adding values... They are just intended to show that the `runner.run` section is going to open, and the user should look for `stopOnFirstFail` or `pageLoadTimeout` in the parameter description.

File: docs/articles/documentation/reference/command-line-interface.md [hunk 1]
```diff
@@ -719,6 +720,21 @@ testcafe ie my-tests --page-load-timeout 0
 
 *Related configuration file property*: [pageLoadTimeout](configuration-file.md#pageloadtimeout).
 
+### --browser-init-timeout \<ms\>
+
+Specifies the time (in milliseconds) for every of the specified browsers to connect and report that it is ready to test. If one or more of the browsers fail to connect within the specified period, an error is thrown.
```
Comment (VasilyStrelyaev) [inline docs/articles/documentation/reference/command-line-interface.md:725]: > every of the specified browsers

Can't this be simplified to *every specified browser*?

Comment (VasilyStrelyaev) [reply docs/articles/documentation/reference/command-line-interface.md:725]: > one or more of the browsers

*one or more browsers* is also legit, isn't it?

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md [hunk 2]
```diff
@@ -11,11 +11,11 @@ A connection to a [remote browser](../../../guides/concepts/browsers.md#browsers
 
 Use this object to run tests on a remote or a mobile device.
 
-`BrowserConnection` is created with the [testCafe.createBrowserConnection](../testcafe/createbrowserconnection.md) function.
+You can create a new `BrowserConnection` with the [testCafe.createBrowserConnection](../testcafe/createbrowserconnection.md) function.
 
-The [browserConnection.url](url.md) property returns a URL the remote browser should visit in order to connect to the [TestCafe server instance](../testcafe/README.md).
+The [browserConnection.url](url.md) property returns a URL a remote browser can visit to connect to the [TestCafe server instance](../testcafe/README.md).
```
Comment (VasilyStrelyaev) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]: > a URL a remote browser can visit to...

Pretty hard to parse grammar here (although it's correct)

*a URL that a remote browser can visit* is easier to understand IMO

**Review by VasilyStrelyaev — CHANGES_REQUESTED** [review]

File: docs/articles/documentation/reference/command-line-interface.md [hunk 3]
```diff
@@ -719,6 +720,21 @@ testcafe ie my-tests --page-load-timeout 0
...
+
+```sh
+testcafe chrome my-tests --browser-init-timeout 40000
```
Comment (wentwrong) [inline docs/articles/documentation/reference/command-line-interface.md:728]: It can be helpful to describe this example a little bit. Like here we changed the default timeout of 2 minutes for Chrome, so TestCafe will dedicate 40 seconds for the browser to establish a connection. Also I think it would be better to use the timeout greater than 40 sec, because probably the common user scenario would be increasing the default timeout (e.g. 180000 ms).

File: docs/articles/documentation/reference/configuration-file.md [hunk 4]
```diff
@@ -709,6 +710,24 @@ See the command line [--page-load-timeout](command-line-interface.md#--page-load
 *CLI*: [--page-load-timeout](command-line-interface.md#--page-load-timeout-ms)  
 *API*: [runner.run({ pageLoadTimeout })](testcafe-api/runner/run.md)
 
+## browserInitTimeout
+
+Specifies the time (in milliseconds) for every of the specified browsers to connect and report that it is ready to test. If one or more of the browsers fail to connect within the specified period, an error is thrown.
```
Comment (wentwrong) [inline docs/articles/documentation/reference/configuration-file.md:715]: Perhaps we should clarify where the browsers will connect to (to the TestCafe server)?

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
+---
+layout: docs
+title: BrowserConnection.opened Event
+permalink: /documentation/reference/testcafe-api/browserconnection/opened.html
+---
+# BrowserConnection.opened Event
+
+Fires when a browser launched by a [browser provider](../../../guides/extend-testcafe/browser-provider-plugin.md#implement-the-browser-provider) has connected to TestCafe and opened a test page.
```
Comment (wentwrong) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:8]: > and opened a test page

I'm not sure about that. It's better to say that a browser is ready to run tests or smth like that.

**Review by wentwrong — CHANGES_REQUESTED** [review]

File: docs/articles/documentation/reference/configuration-file.md (hunk 4 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/configuration-file.md:715]: Is it correct to say that browsers connect to the [TestCafe Object](https://devexpress.github.io/testcafe/documentation/reference/testcafe-api/testcafe/)

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 2 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]: What about:
> The browserConnection.url property returns a URL that remote browsers can visit to connect to the [TestCafe server instance](../testcafe/README.md).

File: docs/articles/documentation/reference/command-line-interface.md (hunk 1 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/command-line-interface.md:725]: I thought about it. 

> Time (in milliseconds) for **all** browsers to connect to TestCafe and report that they are ready to test. If one or more of the browsers fail to connect within the specified period, an error is thrown.

OR

> Time (in milliseconds) for every browser to connect to TestCafe and report that it is ready to test. If one or more of the browsers fail to connect within the specified period, an error is thrown.

Is it obvious that we mean `all browsers you specified`, not `all browsers in the world`?

File: docs/articles/documentation/reference/command-line-interface.md (hunk 3 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/command-line-interface.md:728]: CLI:
>In this example, the timeout for a local browser is increased. The browser has three minutes to connect before TestCafe throws an error.

Config:
> In this example, the timeout for local and remote browsers is three minutes. In this run, all browsers have to connect within this time before TestCafe throws an error.

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md (hunk 5 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:8]: > ...has connected to TestCafe and reported that it is ready to run tests.

File: docs/articles/documentation/reference/configuration-file.md [hunk 6]
```diff
@@ -709,6 +710,26 @@ See the command line [--page-load-timeout](command-line-interface.md#--page-load
...
+
+Time (in milliseconds) for browsers to connect to TestCafe and report that they are ready to test. If one or more browsers fail to connect within the specified period, an error is thrown.
+
+```json
+{
+    "browserInitTimeout": 180000
+}
+```
+
+In this example, the timeout for all browsers is three minutes. In this run, local and remote browsers have to connect within this time before TestCafe throws an error.
```
Comment (VasilyStrelyaev) [inline docs/articles/documentation/reference/configuration-file.md:723]: > In this run, local and remote browsers have to connect within this time

I'm not sure about `have to` here. Like browsers have no other choice but to connect.

**Review by VasilyStrelyaev — APPROVED** [review]
just one minor issue

File: docs/articles/documentation/reference/configuration-file.md (hunk 6 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/configuration-file.md:723]: > If all local and remote browsers don't connect within this time, TestCafe throws an error.

File: docs/articles/documentation/reference/configuration-file.md (hunk 4 above)
Comment (wentwrong) [reply docs/articles/documentation/reference/configuration-file.md]: Yes, I think it's correct.

File: docs/articles/documentation/reference/command-line-interface.md (hunk 3 above)
Comment (wentwrong) [reply docs/articles/documentation/reference/command-line-interface.md]: It's good, but I think in the first one (CLI), we can change "a local browser" to "Chrome" to be more specific. Because "--browser-init-timeout" will change the timeout both for local and remote browsers.

Comment (DIRECTcut) [reply docs/articles/documentation/reference/command-line-interface.md]: >In this example, browser timeout for Chrome is increased. The browser has three minutes to initialize before TestCafe throws an error.

**Review by wentwrong — APPROVED** [review]

File: docs/articles/documentation/reference/testcafe-api/runner/run.md [hunk 7]
```diff
@@ -85,11 +87,11 @@ finally {
 If a browser stops responding while it executes tests, TestCafe restarts the browser and reruns the current test in a new browser instance.
 If the same problem occurs with this test two more times, the test run finishes and an error is thrown.
 
-> When you use a [LiveModeRunner](../livemoderunner.md), you can call the `runner.run` method only once. In rare cases when you need multiple live mode sessions to run in parallel, you can create several [TestCafe server instances](../testcafe/README.md).
+> When you use a [LiveModeRunner](../livemoderunner.md), you can call the `runner.run` method only once. In rare cases when you need multiple live mode sessions to run in parallel, you can create multiple [TestCafe server instances](../testcafe/README.md).
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/runner/run.md:90]:
```suggestion
> When you use a [LiveModeRunner](../livemoderunner.md), can call the `runner.run` method only once. If you want to run multiple live mode sessions simultaneously, you can create multiple [TestCafe server instances](../testcafe/README.md).
```

File: docs/articles/documentation/reference/testcafe-api/runner/run.md [hunk 8]
```diff
@@ -11,7 +11,7 @@ Runs tests according to the current configuration. Returns the number of failed
 async run(options) → Promise<Number>
 ```
 
-Before TestCafe runs tests, it reads settings from the `.testcaferc.json` [configuration file](../../configuration-file.md) if this file exists. Then it applies settings specified in the programming API. API settings override values from the configuration file in case they differ. TestCafe prints information about every overridden property in the console.
+Before TestCafe runs tests, it reads settings from the `.testcaferc.json` [configuration file](../../configuration-file.md) if this file exists. Then it applies settings specified in the programming API. API settings override values from the configuration file if they differ. TestCafe prints information about every overridden property in the console.
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/runner/run.md:14]:
```suggestion
Before TestCafe runs tests, it reads settings from the `.testcaferc.json` [configuration file](../../configuration-file.md) if this file exists and applies settings specified in the programming API. API settings override values from the configuration file. TestCafe prints information about every overridden property in the console.
```

File: docs/articles/documentation/reference/testcafe-api/runner/run.md [hunk 9]
```diff
@@ -11,7 +11,7 @@ Runs tests according to the current configuration. Returns the number of failed
...
 
 > Important! Make sure to keep the browser tab that runs tests active. Do not minimize the browser window.
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/runner/run.md:16]:
```suggestion
> Important! Make sure the browser tab that runs tests stays active. Do not minimize the browser window.
```

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
+---
+layout: docs
+title: BrowserConnection.opened Event
+permalink: /documentation/reference/testcafe-api/browserconnection/opened.html
+---
+# BrowserConnection.opened Event
+
+Fires when a browser launched by a [browser provider](../../../guides/extend-testcafe/browser-provider-plugin.md#implement-the-browser-provider) has connected to TestCafe and reported that it is ready to run tests.
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:8]:
```suggestion
Fires when a browser launched by a [browser provider](../../../guides/extend-testcafe/browser-provider-plugin.md#implement-the-browser-provider) connects to TestCafe and reports that it is ready to run tests.
```

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md [hunk 11]
```diff
@@ -11,11 +11,11 @@ A connection to a [remote browser](../../../guides/concepts/browsers.md#browsers
 
 Use this object to run tests on a remote or a mobile device.
 
-`BrowserConnection` is created with the [testCafe.createBrowserConnection](../testcafe/createbrowserconnection.md) function.
+You can create a new `BrowserConnection` with the [testCafe.createBrowserConnection](../testcafe/createbrowserconnection.md) function.
 
-The [browserConnection.url](url.md) property returns a URL the remote browser should visit in order to connect to the [TestCafe server instance](../testcafe/README.md).
+The [browserConnection.url](url.md) property returns a URL that remote browsers can visit to connect to the [TestCafe server instance](../testcafe/README.md).
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]: "browsers can visit to connect to" This does not make sense

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md [hunk 12]
```diff
@@ -29,7 +29,7 @@ const remoteConnection = await testcafe.createBrowserConnection();
 // Outputs remoteConnection.url so that it can be visited from the remote browser.
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:29]:
```suggestion
// Outputs the remoteConnection.url to allow access from the remote browser.
```

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+```js
+browserConnection.once('opened', callback)
+```
+
+If every remote browser connection does not emit this event within **6** minutes, TestCafe throws an error. You can configure this timeout with the [browserInitTimeout](../../configuration-file.md#browserinittimeout) configuration file property.
```
Comment (dirk-pieterse) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:14]: "does not emit this event" Why do you use **emit**?

**Review by dirk-pieterse — APPROVED** [review]

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md (hunk 13 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:14]: [All objects that emit events are instances of the EventEmitter class](https://nodejs.org/api/events.html#events_events:~:text=All%20objects%20that%20emit%20events%20are%20instances%20of%20the%20EventEmitter%20class.)

Therefore, `BrowserConnection` is an event emitter too. 

Maybe that's too technical. If it sounds better, we could go with:

> If every remote browser connection does not fire this event within **6** minutes...

OR

> If every remote browser connection does not call this event within **6** minutes,

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 11 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]: I meant:

> ... a **URL** that remote browsers can visit to connect to...

Maybe that's hard to parse.  What about.

> Remote browsers can visit a URL returned by the `browserConnection.url` property  to connect to a [TestCafe server instance](../testcafe/README.md).

OR

> The `browserConnection.url` property returns a URL. Remote browsers can visit this URL to connect to a TestCafe server instance.

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 12 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:29]: This sentence implies `any remote browser`. Can we change it to:

> Outputs the `remoteConnection.url` to allow access from a remote browser.

File: docs/articles/documentation/reference/testcafe-api/runner/run.md (hunk 7 above)
Comment (DIRECTcut) [reply docs/articles/documentation/reference/testcafe-api/runner/run.md:90]: Can we simplify:

> When you use a LiveModeRunner, call the `runner.run` method **once**. To run multiple live mode sessions simultaneously, create multiple TestCafe server instances.

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 11 above)
Comment (dirk-pieterse) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]: The browserConnection.url property returns a URL. Remote browsers can use this URL to connect to a TestCafe server instance.

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 12 above)
Comment (dirk-pieterse) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:29]: Yes

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md (hunk 13 above)
Comment (dirk-pieterse) [reply docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:14]: If every remote browser connection does not fire this event within 6 minutes...

File: docs/articles/documentation/reference/testcafe-api/runner/run.md (hunk 7 above)
Comment (dirk-pieterse) [reply docs/articles/documentation/reference/testcafe-api/runner/run.md:90]: Ok

**Review by dirk-pieterse — APPROVED** [review]

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 11 above)
Comment (DIRECTcut) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:16]:
```suggestion
The [browserConnection.url](url.md) property returns a URL. Remote browsers can use this URL to connect to a [TestCafe server instance](../testcafe/README.md).
```

File: docs/articles/documentation/reference/testcafe-api/browserconnection/README.md (hunk 12 above)
Comment (DIRECTcut) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/README.md:29]:
```suggestion
// Outputs the remoteConnection.url to allow access from a remote browser.
```

File: docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+---
+# BrowserConnection.opened Event
+
+Fires when a browser launched by a [browser provider](../../../guides/extend-testcafe/browser-provider-plugin.md#implement-the-browser-provider) connects to TestCafe and reports that it is ready to run tests.
+
+```js
+browserConnection.once('opened', callback)
+```
+
+If every remote browser connection does not emit this event within **6** minutes, TestCafe throws an error. You can configure this timeout with the [browserInitTimeout](../../configuration-file.md#browserinittimeout) configuration file property.
```
Comment (DIRECTcut) [inline docs/articles/documentation/reference/testcafe-api/browserconnection/opened.md:14]:
```suggestion
If every remote browser connection does not fire this event within **6** minutes, TestCafe throws an error. You can configure this timeout with the [browserInitTimeout](../../configuration-file.md#browserinittimeout) configuration file property.
```

File: docs/articles/documentation/reference/testcafe-api/runner/run.md (hunk 7 above)
Comment (DIRECTcut) [inline docs/articles/documentation/reference/testcafe-api/runner/run.md:90]:
```suggestion
> When you use a [LiveModeRunner](../livemoderunner.md), call the `runner.run` method once. To run multiple live mode sessions simultaneously, create multiple [TestCafe server instances](../testcafe/README.md).
```
